src/python/WMCore/MicroService/MSPileup/MSPileupTasks.py

```diff
@@ -0,0 +1,337 @@
+           - if there is none, then a new rule needs to be created. First, check whether the RSE has enough space available for that
+             - if it does, create the rule and append the rule id to rulesList
+             - otherwise, make a log record saying that there is not enough space
+        4. once all the relevant rules have been created, persist an up-to-date version of the pileup data structure in MongoDB
+
+        """
+        spec = {'active': True}
+        docs = self.mgr.getPileup(spec)
+        taskSpec = self.getTaskSpec()
+        taskSpec['task'] = 'active'
+        taskSpec['marginSpace'] = marginSpace
+        asyncio.run(performTasks('active', docs, taskSpec))
+
+    def getTaskSpec(self):
+        """Return task spec"""
+        spec = {'manager': self.mgr, 'rucioClient': self.rucio, 'rucioAccount': self.rucioAccount}
+        return spec
+
+
+def getRucioClient(account, rucioHost='http://cms-rucio.cern.ch', authHost='https://cms-rucio-auth.cern.ch'):
+    """
+    Helper function to provide Rucio client
+    :param account: rucio account to use
+    :param rucioHost: rucio host to use
+    :param authHost: rucio auth host to use
+    :return: rucio client
+    """
+    creds = {"client_cert": os.getenv("X509_USER_CERT", "Unknown"),
+             "client_key": os.getenv("X509_USER_KEY", "Unknown")}
+
+    client = Client(rucio_host=rucioHost,
+                    auth_host=authHost,
+                    account=account,
+                    ca_cert=False,
+                    auth_type="x509",
+                    timeout=30,
+                    creds=creds)
+    return client
+
+
+def rucioArgs(account):
+    """Return relevant rucio arguments"""
+    kwargs = {'groupping': 'ALL', 'account': account,
+              'lifeTime': None, 'locked': False, 'notify': 'N',
+              'purge_replicas': False, 'ignore_availability': False,
+              'ask_approval': False, 'asynchrounous': True, 'priority': 3}
+    return kwargs
+
+
+async def monitoringTask(doc, spec):
+    """
+    Perform single monitoring task over provided MSPileup document
+
+        2. For each rule id in rulesList:
+           - query Rucio for that rule id and fetch its state (e.g.: afd122143kjmdskj)
+           - if state=OK, log that the rule has been satisfied and add that RSE to the currentRSEs (unique)
+           - otherwise, calculate the rule completion based on the 3 locks_* field
+        3. now that all the known rules have been inspected, persist the up-to-date pileup doc in MongoDB
+
+    :param mgr: MSPileupData manager
+    :param doc: MSPileup document
+    :param spec: task spec dict
+    """
+    mgr = spec['manager']
+    rucioClient = spec['rucioClient']
+    logger = spec.get('logger', getMSLogger(False))
+    # get list of existentn rule ids
+    pname = doc['pileupName']
+    inputArgs = {"scope": "cms", "name": pname}
+    rules = [r for r in rucioClient.list_replication_rules(inputArgs)]
```

Comment:
This is very interesting question since:
- I called a single rucio API which returns a list of rules, where each individual dict contain rule ID
- you are asked me to make N calls to rucio, one for each individual ruie ID to get their info

My question is which approach is more appropriate and how rucio output differ in both cases.

Reply:
Valentin, as we discussed over Zoom. Let's take the approach that requires the least number of HTTP calls (if data volume returned is still small). So, please keep the current implementation (but add the account filter to it).

Reply:
I added account into input args dict